### DISCUSSION TITLE
Defect Database

### DISCUSSION THREAD

**Discussion by fupsi42** [netbox-community/netbox#10393] [Q&A]
hi all,

when trying to update from 3.2.6 to 3.3.2 I received errors. I went back to 3.2.6 through a snapshot of the system and checked my entries there. I have 4 cables which I can't delete with 8 connections (which I don't see). I get the following error:

<class 'dcim.models.cables.Cable.DoesNotExist'>

Cable matching query does not exist.

Python version: 3.9.2
NetBox version: 3.2.6

On further checking it turns out that all my devices are no longer in the database. There are 116 interfaces according to the balloon in the bar and according to the api, but the list remains empty.
Also, since some update, all the associated devices have been deleted. According to the history, the last time I made changes to devices was in June.

Everything under Virtualisation and IPAM is still correct.
Is there a way to delete the devices and connection section in the database so that I can continue to update cleanly? The entries there do not bother me when they are gone.

Many thanks in advance!

**Comment by candlerb** [comment]
So you have device interfaces in the database (table `dcim_interface`), which are not associated with any device??

Try this query in postgres:

```
netbox=# select id,name,device_id from dcim_interface;
```

and look for devices with the given device_id (`select id,name from dcim_device where id=123;`)

However, that should not be possible, because there should be a foreign key constraint which prevents it. 

```
netbox=# \d dcim_interface;
...
    "dcim_interface_device_id_359c6115_fk" FOREIGN KEY (device_id) REFERENCES dcim_device(id) DEFERRABLE INITIALLY DEFERRED
```

If you're missing that constraint, then your database is badly corrupted - perhaps some migrations were not correctly applied during updates from older versions of Netbox.

You'll need a SQL expert to fix this, together with a freshly installed Netbox of the same version.  You could either compare the schemas manually and apply fixups by hand; or export the contents of the tables of one system into the new system.  But you'll really need to know what you're doing with SQL to do that.

**Reply by fupsi42** [reply]
I have executed the query and the interfaces are there. 
However, the foreign-key constraints are missing.

I fear bad things...

If there is no easy way to delete the broken data, is it possible to export at least the IPAM and VM data?

**Reply by candlerb** [reply]
In principle it's easy to delete the broken data: just delete any interface rows where the device_id points to a non-existent device. You'll also need to delete or set-to-null any ipam_ipaddress which points at the device (check assigned_object_id and assigned_object_type_id; the latter is a reference into django_content_type).

Once that's done, you can apply the missing foreign key constraint.  But if your database is messed up in this way, it could well be messed up in others unfortunately.  No shortcuts to fixing this I'm afraid.

You *can* kind-of export the IPAM and VM data, but unfortunately not in a form that's useful for re-importing into another instance of Netbox.
